stream-chat-android-client/src/main/java/io/getstream/chat/android/client/utils/FiniteStateMachine.kt

```diff
@@ -0,0 +1,96 @@
+package io.getstream.chat.android.client.utils
+
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.reflect.KClass
+
+internal typealias StateFunction<S, E> = FiniteStateMachine<S, E>.(S, E) -> S
+
+internal class FiniteStateMachine<S : State, E : Event>(
```

Comment:
Is this completely our own implementation for this, or is it based on something?

Reply:
It's completely mine implementation. Based on my previous experience and made a little less simple (usually in FSM and in graph there is some collection of side effects)